File: book/lib/services/kernel_service.dart
import '../models/kernel.dart';
import 'api_client.dart';

class KernelService {
  final ApiClient _api;

  KernelService({ApiClient? api}) : _api = api ?? apiClient;

  Future<List<Kernel>> list() async {
    try {
      final response = await _api.getList('/api/kernels');
      return response.map((k) => Kernel.fromJson(k as Map<String, dynamic>)).toList();
    } catch (e) {
      return [];
    }
  }

  Future<Kernel> create(String name, {String? notebookId}) async {
    final response = await _api.post('/api/kernels', {
      'name': name,
      if (notebookId != null) 'notebook_id': notebookId,
    });
    return Kernel.fromJson(response);
  }

  Future<Kernel> get(String kernelId) async {
    final response = await _api.get('/api/kernels/$kernelId');
    return Kernel.fromJson(response);
  }

  Future<KernelStatus> getStatus(String kernelId) async {
    final response = await _api.get('/api/kernels/$kernelId/status');
    return KernelStatus.values.firstWhere(
      (s) => s.name == response['status'],
      orElse: () => KernelStatus.dead,
    );
  }

  Future<void> interrupt(String kernelId) async {
    await _api.post('/api/kernels/$kernelId/interrupt', {});
  }

  Future<Kernel> restart(String kernelId) async {
    final response = await _api.post('/api/kernels/$kernelId/restart', {});
    return Kernel.fromJson(response);
  }

  Future<void> shutdown(String kernelId) async {
    await _api.delete('/api/kernels/$kernelId');
  }

  Future<List<dynamic>?> getVariables(String kernelId) async {
    try {
      final response = await _api.get('/api/kernels/$kernelId/variables');
      return response['variables'] as List<dynamic>?;
    } catch (e) {
      return null;
    }
  }

  Future<Map<String, dynamic>?> complete(String kernelId, String code, int cursorPos) async {
    try {
      final response = await _api.post('/api/kernels/$kernelId/complete', {
        'code': code,
        'cursor_pos': cursorPos,
      });
      return response;
    } catch (e) {
      return null;
    }
  }

  Future<Map<String, dynamic>?> inspect(String kernelId, String code, int cursorPos) async {
    try {
      final response = await _api.post('/api/kernels/$kernelId/inspect', {
        'code': code,
        'cursor_pos': cursorPos,
      });
      return response;
    } catch (e) {
      return null;
    }
  }
}

final kernelService = KernelService();

File: notebook/src/services/kernelService.ts
// Kernel Service - Manage IPython kernels

import apiClient from "./apiClient";

export interface Kernel {
  id: string;
  name: string;
  status: KernelStatus;
  notebookId?: string;
  lastActivity: Date;
  executionCount: number;
  connections: number;
}

export enum KernelStatus {
  IDLE = "idle",
  BUSY = "busy",
  STARTING = "starting",
  DEAD = "dead",
  RESTARTING = "restarting",
}

export interface KernelSpec {
  name: string;
  displayName: string;
  language: string;
  argv: string[];
  interruptMode?: string;
}

interface KernelResponse {
  id: string;
  name: string;
  status: string;
  notebook_id?: string;
  last_activity?: string;
  created_at?: string;
  execution_count?: number;
  connections?: number;
}

class KernelService {
  private parseKernel(data: KernelResponse): Kernel {
    return {
      id: data.id,
      name: data.name,
      status: data.status as KernelStatus,
      notebookId: data.notebook_id,
      lastActivity: new Date(data.last_activity || data.created_at || Date.now()),
      executionCount: data.execution_count || 0,
      connections: data.connections || 0,
    };
  }

  async list(): Promise<Kernel[]> {
    const response = await apiClient.get<KernelResponse[]>("/api/kernels");
    return response.map(this.parseKernel);
  }

  async get(id: string): Promise<Kernel> {
    const response = await apiClient.get<KernelResponse>(`/api/kernels/${id}`);
    return this.parseKernel(response);
  }

  async create(name: string = "python3", notebookId?: string): Promise<Kernel> {
    const response = await apiClient.post<KernelResponse>("/api/kernels", {
      name,
      notebook_id: notebookId,
    });
    return this.parseKernel(response);
  }

  async restart(id: string): Promise<Kernel> {
    const response = await apiClient.post<KernelResponse>(`/api/kernels/${id}/restart`, {});
    return this.parseKernel(response);
  }

  async interrupt(id: string): Promise<void> {
    await apiClient.post(`/api/kernels/${id}/interrupt`, {});
  }

  async shutdown(id: string): Promise<void> {
    await apiClient.delete(`/api/kernels/${id}`);
  }

  async shutdownAll(): Promise<void> {
    const kernels = await this.list();
    await Promise.all(kernels.map((k) => this.shutdown(k.id)));
  }

  async getSpecs(): Promise<Record<string, KernelSpec>> {
    return apiClient.get<Record<string, KernelSpec>>("/api/kernels/specs");
  }

  // Get or create kernel for a notebook
  async getOrCreateForNotebook(notebookId: string): Promise<Kernel> {
    const kernels = await this.list();
    const existing = kernels.find((k) => k.notebookId === notebookId && k.status !== KernelStatus.DEAD);

    if (existing) {
      return existing;
    }

    return this.create("python3", notebookId);
  }
}

export const kernelService = new KernelService();
export default kernelService;
